### PR TITLE
Add Narc/Improved Exile Screen

### PR DESCRIPTION
**Pull request by impostor4291**
Narc:https://discord.com/channels/1094344790910455908/1309925307163086948

### CODE REVIEW

File: Roles/AddOns/Impostor/Narc.cs
Comment (MargaretTheFool) [inline Roles/AddOns/Impostor/Narc.cs:1]: make the id for narc 31100

File: Modules/CustomRolesHelper.cs [hunk 1]
```diff
@@ -362,7 +362,9 @@ or CustomRoles.Recruit
             or CustomRoles.Infected
             or CustomRoles.Contagious
             or CustomRoles.Rascal
-            or CustomRoles.Soulless;
+            or CustomRoles.Soulless
+            or CustomRoles.Admired
```
Comment (NikoCat233) [inline Modules/CustomRolesHelper.cs:366]: admired doesnt work as converted

File: Modules/GameState.cs [hunk 2]
```diff
@@ -107,6 +107,10 @@ public void SetMainRole(CustomRoles role)
         {
             countTypes = CountTypes.Crew;
         }
+        if (pc.Is(CustomRoles.Narc))
```
Comment (NikoCat233) [inline Modules/GameState.cs:110]: is narc a main role or a sub role?
oc.is(role) is used to check main role i think

File: Patches/IntroPatch.cs [hunk 3]
```diff
@@ -302,7 +302,18 @@ public static bool Prefix(IntroCutscene __instance, ref Il2CppSystem.Collections
...
+            var narcDisplay = new Il2CppSystem.Collections.Generic.List<PlayerControl>();
+            narcDisplay.Add(PlayerControl.LocalPlayer);
+            foreach (var pc in Main.AllAlivePlayerControls)
+            {
+                if (pc != PlayerControl.LocalPlayer)
+                    narcDisplay.Add(pc);
+            }
+            teamToDisplay = narcDisplay;
+        }
+        else if ((role.IsMadmate() || PlayerControl.LocalPlayer.Is(CustomRoles.Madmate)) && !PlayerControl.LocalPlayer.Is(CustomRoles.Narc))
```
Comment (NikoCat233) [inline Patches/IntroPatch.cs:316]: double if narc

File: Patches/IntroPatch.cs [hunk 4]
```diff
@@ -445,12 +456,9 @@ public static void Postfix(IntroCutscene __instance)
                 PlayerControl.LocalPlayer.Data.Role.IntroSound = DestroyableSingleton<HudManager>.Instance.TaskCompleteSound;
                 break;
 
-            case CustomRoles.Sheriff:
-            case CustomRoles.Veteran:
```
Comment (NikoCat233) [inline Patches/IntroPatch.cs:449]: any necessary reason for changing this?

File: Resources/Lang/en_US.json [hunk 5]
```diff
@@ -2644,18 +2648,19 @@
   "ExiledInnocentTargetInOneLine": "{0} was the {1}.\nBut looking back, there's the <color=#8f815e>Innocent</color> counting the money in their hands....\nGG!",
   "ExiledDeath": "{0} was {1}!\nThe Crew has been saved from <color=#ff174f>Armageddon</color>!",
   "ExiledNotDeath": "{0} was the {1}.\nBut they were not <color=#644661>Death</color>...\n<b><color=#ff174f>Death has claimed the souls of the Crew!</color></b>",
-  "IsGood": "{0} was a <color=#8cffff>good guy</color>",
+  "IsGood": "{0} was a <color=#8cffff>Crewmate</color>",
```
Comment (NikoCat233) [inline Resources/Lang/en_US.json:2651]: wtf is this sentence

File: Roles/AddOns/Impostor/Narc.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+namespace TOHE.Roles.AddOns.Impostor;
+
+public class Narc : IAddon
+{
+    private const int Id = 23400;
+    public AddonTypes Type => AddonTypes.Misc;
+
+    public static OptionItem MeetingsNeededForWin;
+
+    public void SetupCustomOption()
```
Comment (NikoCat233) [inline Roles/AddOns/Impostor/Narc.cs:13]: override

File: Roles/Core/CustomRoleManager.cs [hunk 7]
```diff
@@ -218,6 +218,10 @@ public static bool OnCheckMurder(ref PlayerControl killer, ref PlayerControl tar
                         canceled = true;
                         break;
 
+                    case CustomRoles.Narc when target.Is(CustomRoles.Sheriff):
```
Comment (NikoCat233) [inline Roles/Core/CustomRoleManager.cs:221]: should not put this check here

**Review by NikoCat233 — CHANGES_REQUESTED** [review]
quick review, fix bugs

File: Patches/IntroPatch.cs (hunk 4 above)
Comment (PEPPERcula) [reply Patches/IntroPatch.cs:449]: Yeah I agree Veteran, Sheriff, and Nice Guesser intro sound shouldn’t be changed

File: Resources/Lang/en_US.json (hunk 5 above)
Comment (MargaretTheFool) [reply Resources/Lang/en_US.json:2651]: it's when you turn exile mode to team

File: Modules/GameState.cs (hunk 2 above)
Comment (impostor4291) [reply Modules/GameState.cs:110]: > is narc a main role or a sub role? oc.is(role) is used to check main role i think

I thought it was used to check subrole :/

File: Patches/IntroPatch.cs (hunk 4 above)
Comment (HyperAtill) [reply Patches/IntroPatch.cs:449]: When you as an unmodded client get Sheriff, do you see "Crewmate" or "Impostor"?

Comment (impostor4291) [reply Patches/IntroPatch.cs:449]: > When you as an unmodded client get Sheriff, do you see "Crewmate" or "Impostor"?

sheriff?

File: Roles/AddOns/Impostor/Narc.cs (hunk 6 above)
Comment (impostor4291) [reply Roles/AddOns/Impostor/Narc.cs:13]: > override

I don't think it works for add-ons

**Comment by impostor4291** [comment]
> conflicts

how did i not see the missing ":" in en_US.json :omgcry:
there was also an "a" missing in Celebrity.cs...
it should be resolved

**Comment by impostor4291** [comment]
> > conflicts
> 
> how did i not see the missing ":" in en_US.json :omgcry: there was also an "a" missing in Celebrity.cs... it should be resolved

i take my words back

**Comment by impostor4291** [comment]
> > > conflicts
> > 
> > 
> > how did i not see the missing ":" in en_US.json :omgcry: there was also an "a" missing in Celebrity.cs... it should be resolved
> 
> i take my words back

finally

**Comment by NikoCat233** [comment]
Please Reslove conflicts

**Comment by impostor4291** [comment]
屏幕截图 2024-12-04 223704
why is this a conflict?

**Comment by NikoCat233** [comment]
Use ide to resolve my friend
merge dev 2.2.0 into your pr and manually resolve it with ide
dont do it on web

**Pull request closed without merging**